pkg/endpoint/events.go

```diff
@@ -119,6 +120,71 @@ func (e *Endpoint) PolicyRevisionBumpEvent(rev uint64) {
 	}
 }
 
+/// EndpointNoTrackEvent contains all fields necessary to update the NOTRACK rules.
+type EndpointNoTrackEvent struct {
+	ep     *Endpoint
+	annoCB AnnotationsResolverCB
+}
+
+// Handle handles the NOTRACK rule update.
+func (ev *EndpointNoTrackEvent) Handle(res chan interface{}) {
+	e := ev.ep
+	port, err := ev.annoCB(e.K8sNamespace, e.K8sPodName)
+	if err != nil {
+		res <- &EndpointRegenerationResult{
+			err: err,
+		}
+		return
+	}
+
+	// If this endpoint is going away, remove all related NOTRACK rules.
+	if err := e.lockAlive(); err != nil {
+		if e.NoTrackPort != "" {
+			if e.IPv4.IsSet() {
+				iptables.RemoveNoTrackRules(e.IPv4.String(), e.NoTrackPort, false)
+			}
+			if e.IPv6.IsSet() {
+				iptables.RemoveNoTrackRules(e.IPv6.String(), e.NoTrackPort, true)
+			}
+			e.NoTrackPort = ""
+		}
+		res <- &EndpointRegenerationResult{
+			err: nil,
+		}
+		return
+	}
+
+	defer func() {
+		e.unlock()
+	}()
+
+	if port != e.NoTrackPort {
+		log.Debug("Updating NOTRACK rules")
+		if e.IPv4.IsSet() {
+			if e.NoTrackPort != "" {
+				iptables.RemoveNoTrackRules(e.IPv4.String(), e.NoTrackPort, false)
+			}
+			if port != "" {
+				iptables.InstallNoTrackRules(e.IPv4.String(), port, false)
+			}
```

Comment:
Is there any chance that `e.NoTrackPort` could be the same as `port`? In that case, this kind of "delete then add" type logic could cause temporary unintended traffic disruption.

Reply:
> Is there any chance that e.NoTrackPort could be the same as port

This entire block is conditional on `e.NoTrackPort != port`.

But in general, how can we avoid disrupting traffic when manipulate ip table rules?

Reply:
> This entire block is conditional on e.NoTrackPort != port.

:+1: I was jumping between the two rule handling sections and must have overlooked this fact. 

In general, we avoid it by ensuring that there is always a rule in place to handle any ongoing traffic. This can be achieved by dumping existing rules and comparing them to ensure that we only apply deltas on top, or it can be achieved by inserting the new rule in front of the old rule first and then deleting the old rule second.

Reply:
Done. Now only validate port number in the event `Handle`.